news/11314.bugfix.rst

```diff
@@ -0,0 +1 @@
+Ensure that things work properly when using with an old setuptools without _distutils_hack present
```

Comment:
This sounds wrong? AttributeError should happen if _distutils_hack _is_ present, but does not provide a `remove_shim` function.

Reply:
I think I see what is happenning:

1 Install newer version of setuptools with the hack present.
2 Downgrade to older version without it present.

This appears to leave the directory containing it as empty, but deletes the .py file present in the directory.

```
x = __import__("_distutils_hack")
print(x)
print(dir(x))
print(x.__file__)
print(x.__package__)
print(x.__path__)
print(x.__spec__)
```

```
07:51:51  <module '_distutils_hack' (namespace)>
07:51:51  ['__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__']
07:51:51  None
07:51:51  _distutils_hack
07:51:51  _NamespacePath(['/home/jenkins-slave/.jarvis/virtualenvs/executor-0/DataManagement/data-mesh/PR-2993_null/lib/python3.8/site-packages/_distutils_hack'])
07:51:51  ModuleSpec(name='_distutils_hack', loader=<_frozen_importlib_external._NamespaceLoader object at 0x7f9c76a74a90>, submodule_search_locations=_NamespacePath(['/home/jenkins-slave/.jarvis/virtualenvs/executor-0/DataManagement/data-mesh/PR-2993_null/lib/python3.8/site-packages/_distutils_hack']))
```

```
ls -l /home/jenkins-slave/.jarvis/virtualenvs/executor-0/DataManagement/data-mesh/PR-2993_null/lib/python3.8/site-packages/_distutils_hack
```
```
07:51:50  total 4
07:51:50  drwxrwxr-x 2 jenkins-slave jenkins-slave 4096 Jul 28 11:51 __pycache__
```

Reply:
I've seen that before. I regularly have to manually clean up package directories left behind after an uninstall. There is https://github.com/pypa/pip/issues/638, but that was fixed a very long time ago. I've tried replicating your finding - downgrading Setuptools, but I still find that `_distutils_hack` gets removed. Can you figure out what steps are necessary to replicate the issue where `_distutils_hack` is left behind after a downgrade?

Reply:
I'm not sure I'll easily be able to do this as I can only iterate via jenkins and not locally. What I do see is:

```
[2022-07-28T11:51:47.897Z] Installing collected packages: wheel, setuptools
[2022-07-28T11:51:47.897Z]   Found existing installation: setuptools 56.0.0
[2022-07-28T11:51:48.202Z]     Uninstalling setuptools-56.0.0:
[2022-07-28T11:51:48.202Z]       Successfully uninstalled setuptools-56.0.0
[2022-07-28T11:51:48.503Z] Successfully installed setuptools-36.0.1 wheel-0.30.0
```
Which should hopefully show the versions involved.